### ISSUE TITLE
Unidentified "legacy functions"

### ISSUE THREAD

**Issue by sennendoko** [onnx/onnx#59]
I'm having some trouble exporting a model from Pytorch - I don't think it contains any legacy functions, but some code added to Onnx a few days ago seems to think it does:

(Pdb) torch.onnx.export(model, image, 'model.proto', verbose=True)
*** RuntimeError: Tracing of legacy functions is not supported
(Pdb) torch.onnx.export(model, image, 'model.proto')
*** RuntimeError: /mnt/nfs/users/dacart/pytorch/torch/csrc/jit/tracer.h:84: getTracingState: Assertion `var_state == state` failed.

To allow me to debug this, would it be possible to print some informative details of the function(s) that are causing the trouble? Thanks.

**Comment by ezyang** [comment]
Hi @sennendoko, very shortly I'll give you a patch which you can apply which should give you the name of the legacy function in question. The assert failure in the second case is a legit bug.

Is it possible to post the source code that reproduces?

**Comment by sennendoko** [comment]
Thanks! I'd need to extract a minimal example to reproduce the problem, and that would be non-trivial, so would rather wait and see if the legacy function name sheds light on what's happening.

**Comment by ezyang** [comment]
Non-minimal example is also OK, but I understand if you can't :)

**Comment by ezyang** [comment]
Please try this patch:

```
diff --git a/torch/csrc/autograd/python_function.cpp b/torch/csrc/autograd/python_function.cpp
index 9f69000..16cf6c5 100644
--- a/torch/csrc/autograd/python_function.cpp
+++ b/torch/csrc/autograd/python_function.cpp
@@ -630,8 +630,12 @@ static void _trace_create(PyObject* op_obj, THPFunction* bw_obj,
   if (!tracer::isTracing(input_vars))
     return;
 
-  if (!op_obj)
-    throw std::runtime_error("Tracing of legacy functions is not supported");
+  if (!op_obj) {
+    std::ostringstream oss;
+    oss << "Attempted to trace " << Py_TYPE(bw_obj)->tp_name;
+    oss << ", but tracing of legacy functions is not supported";
+    throw std::runtime_error(oss.str());
+  }
 
   auto tracing_state = tracer::getTracingState(input_vars);
   bw_obj->is_traced = true;
```

**Comment by sennendoko** [comment]
Thank you...I appreciate the quick response.

With the patch, it tells me:
*** RuntimeError: Attempted to trace NoGrad, but tracing of legacy functions is not supported

And the call of NoGrad is coming from Variable.detach in torch/autograd/variable.py, which I am indeed calling during a forward pass in my model.

So...should I be looking at avoiding the call of "detach" in my model, or would you be able to either make NoGrad non-legacy or avoid calling it from "detach"?

**Comment by ezyang** [comment]
We should probably refactor NoGrad to be non-legacy / handle this case correctly, but for now, I'd recommend removing the `detach()` call; it won't affect the dump (because detach only affects backwards.)

**Comment by sennendoko** [comment]
Makes sense...so, I have replaced calls in my code of the form "y = x.detach()" by just "y = x" (I also tried "y = Variable(x.data)". But now in either case, I get

(Pdb) torch.onnx.export(model, image, 'model.proto', verbose=True)
*** RuntimeError: untraced buffer

which appears to come from getBufferTrace in torch/csrc/jit/tracer.h, which is called (twice) from torch/csrc/autograd/functions/onnx/batch_normalization.cpp -- and I do have some nn.BatchNorm3d modules in my model. Suggestions welcome once more, and thanks again.

**Comment by ezyang** [comment]
Assuming this isn't a bug in ONNX (which it might be), this error means that for whatever reason, your batchnorm modules aren't being detected for getting their parameters/buffers. To make sure this detection works, you must put all of your batchnorms as members of a module (which itself is a member, perhaps recursively, of the root module you're exporting.)
